File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        totmax=float('-inf')
        curmax=1
        curmin=1
        for i in range(len(nums)):
            temp=curmax*nums[i]
            curmax=max(curmax*nums[i],nums[i],curmin*nums[i])
            curmin=min(temp,nums[i],curmin*nums[i])
            totmax=max(curmax,totmax)
        return totmax
        

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int psum=1;
        int ssum=1;
        int N=nums.length;
        int totsum=Integer.MIN_VALUE;
        for(int i=0;i<N;i++){
            if(psum==0) psum=1;
            if(ssum==0) ssum=1;
            psum=psum*nums[i];
            ssum=ssum*nums[N-i-1];
            totsum=Math.max(Math.max(psum,ssum),totsum);
        }
        return totsum;
    }
}